javascript/1268-search-suggestions-system.js

```diff
@@ -0,0 +1,46 @@
+/**
+ * Binary Search
+ * 
+ * Time O(n*log(n) + m*n) | Space O(m)
+ * https://leetcode.com/problems/search-suggestions-system/description/
+ * @param {string[]} products
+ * @param {string} searchWord
+ * @return {string[][]}
+ */
+var suggestedProducts = function(products, searchWord) {
```

Comment:
Suggestion: Break down function to match O() complexity

```js
products.sort(sortAsc); /* Time O(N * log(N) */
search(...);		/* Time O(......) */
```

Reply:
Suggestion: Look at Trie

```js
/**
 * DFS - Trie
 * Time O(N * M) | Space O(N)
 * https://leetcode.com/problems/search-suggestions-system/
 * @param {string[]} products
 * @param {string} searchWord
 * @return {string[][]}
 */
var suggestedProducts = (products, searchWord) => new Trie()
    .buildTrie(products)
    .searchWord(searchWord);

class Node {
    constructor () {
        this.children = new Map();
        this.isWord = false;
    }
};

class Trie {
    constructor () {
        this.root = new Node();
    }

    buildTrie (products) {
        for (const word of products) {
            this.insert(word);
        }

        return this;
    }
    
    insert (word, { root: node } = this) {        
        for (const char of word.split('')) {
            const child = (node.children.get(char) ?? new Node());
            
            node.children.set(char, child);

            node = child;
        }

        node.isWord = true;
    }
    
    searchWord (searchWord, buffer = [], suggestions = []) {
        for (const char of searchWord.split('')) {
            const prefix = this.getPrefix(buffer, char);
            const words = this.getSuggestions(prefix);

            suggestions.push(words);
        }
        
        return suggestions;
    }

    getPrefix (buffer, char) {
        buffer.push(char);

        return buffer.join('');
    }
    
    getSuggestions (prefix, words = []) {
        const node = this.getPrefixNode(prefix);

        const isInvalidPrefix = (node === null);
        if (isInvalidPrefix) return words
        
        return this.search(node, prefix, words);
    }

    getPrefixNode (prefix, { root: node } = this) {
        for (const char of prefix.split('')) {
            const child = (node.children.get(char) ?? null);

            const isLeafNode = (child === null);
            if (isLeafNode) return null;

            node = child;
        }

        return node;
    }

    search (node, word, words) {
        const isBaseCase = (words.length === 3);
        if (isBaseCase) return words;

        if (node.isWord) words.push(word);

        return this.dfs(node, word, words);
    }

    dfs (node, word, words) {
        for (const char of this.getChars()) {
            const child = (node.children.get(char) ?? null);

            const isLeafNode = (child === null);
            if (isLeafNode) continue;
        
            this.search(child, (word + char), words);
        }

        return words;
    }

    getChars () {
        return new Array(26).fill()
            .map((_, index) => String.fromCharCode((index + 97)));
    }
};
```

Reply:
Don't you think it's too much? I mean, I might be intimidating for some people. NeetCode solved it with binary-search that's how I solved it too. It was more concise that way. But, if you want, I can add this solution in addition to the binary-search one.

Reply:
Trie is typically a better solution since for searching, suggestions, and prefix against words.
Also the complexity is better.

Reply:
Ok, I'll implement that.

Reply:
@aakhtar3, I added the trie solution. But I kept the binary-search solution for reference.